org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/format/XMLFormatterDocumentNew.java

```diff
@@ -0,0 +1,559 @@
+import org.eclipse.lemminx.dom.DOMProcessingInstruction;
+import org.eclipse.lemminx.dom.DOMText;
+import org.eclipse.lemminx.services.extensions.format.IFormatterParticipant;
+import org.eclipse.lemminx.settings.SharedSettings;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.TextEdit;
+import org.eclipse.lsp4j.jsonrpc.CancelChecker;
+import org.w3c.dom.Node;
+import org.w3c.dom.Text;
+
+/**
+ * Experimental XML formatter which generates several text edit to remove, add,
+ * update spaces / indent.
+ * 
+ * @author Angelo ZERR
+ *
+ */
+public class XMLFormatterDocumentNew {
+
+	private static final Logger LOGGER = Logger.getLogger(XMLFormatterDocumentNew.class.getName());
+
+	private static final String XML_SPACE_ATTR = "xml:space";
+
+	private static final String XML_SPACE_ATTR_DEFAULT = "default";
+
+	private static final String XML_SPACE_ATTR_PRESERVE = "preserve";
+
+	private final DOMDocument xmlDocument;
+	private final TextDocument textDocument;
+	private final String lineDelimiter;
+	private final SharedSettings sharedSettings;
+
+	private final DOMProcessingInstructionFormatter processingInstructionFormatter;
+
+	private final DOMDocTypeFormatter docTypeFormatter;
+
+	private final DOMElementFormatter elementFormatter;
+
+	private final DOMAttributeFormatter attributeFormatter;
+
+	private final DOMTextFormatter textFormatter;
+
+	private final Collection<IFormatterParticipant> formatterParticipants;
+
+	private int startOffset = -1;
+	private int endOffset = -1;
+
+	private CancelChecker cancelChecker;
+
+	/**
+	 * XML formatter document.
+	 */
+	public XMLFormatterDocumentNew(DOMDocument xmlDocument, Range range, SharedSettings sharedSettings,
+			Collection<IFormatterParticipant> formatterParticipants) {
+		this.xmlDocument = xmlDocument;
+		this.textDocument = xmlDocument.getTextDocument();
+		this.lineDelimiter = computeLineDelimiter(textDocument);
+		if (range != null) {
+			try {
+				startOffset = textDocument.offsetAt(range.getStart());
+				endOffset = textDocument.offsetAt(range.getEnd());
+			} catch (BadLocationException e) {
+				LOGGER.log(Level.SEVERE, e.getMessage(), e);
+			}
+		}
+		this.sharedSettings = sharedSettings;
+		this.formatterParticipants = formatterParticipants;
+		this.docTypeFormatter = new DOMDocTypeFormatter(this);
+		this.attributeFormatter = new DOMAttributeFormatter(this);
+		this.elementFormatter = new DOMElementFormatter(this, attributeFormatter);
+		this.processingInstructionFormatter = new DOMProcessingInstructionFormatter(this, attributeFormatter);
+		this.textFormatter = new DOMTextFormatter(this);
+
+	}
+
+	private static String computeLineDelimiter(TextDocument textDocument) {
+		try {
+			return textDocument.lineDelimiter(0);
+		} catch (BadLocationException e) {
+			LOGGER.log(Level.SEVERE, e.getMessage(), e);
+		}
+		return System.lineSeparator();
+	}
+
+	/**
+	 * Returns a List containing a single TextEdit, containing the newly formatted

```

Comment:
It looks like sometimes multiple `TextEdit` are returned. Update the documentation to reflect this.

Reply:
good catch, fixed.